frost/src/spatial.rs

```diff
@@ -0,0 +1,476 @@
+use crate::{Error, FrostLatLonElev, FrostLocation, FrostObs};
+use chrono::prelude::*;
+use rove::data_switch::{SpatialCache, Timestamp};
+use rove::pb::GeoPoint;
+
+fn extract_metadata(
+    mut header: serde_json::Value,
+    time: DateTime<Utc>,
+) -> Result<FrostLatLonElev, Error> {
+    let location = header
+        .get_mut("extra")
+        .ok_or(Error::FindLocation(
+            "couldn't find extra in header".to_string(),
+        ))?
+        .take()
+        .get_mut("station")
+        .ok_or(Error::FindLocation(
+            "couldn't find station field in extra".to_string(),
+        ))?
+        .get_mut("location")
+        .ok_or(Error::FindLocation(
+            "couldn't find location field in station".to_string(),
+        ))?
+        .take();
+
+    let loc = serde_json::from_value::<Vec<FrostLocation>>(location)?;
+
+    let lat_lon_elev = loc
+        .into_iter()
+        .find(|l| time > l.from && time < l.to)
+        .ok_or(Error::FindLocation(
+            "couldn't find relevant location for this observation".to_string(),
+        ))?
+        .value;
+
+    Ok(lat_lon_elev)
+}
+
+fn extract_data(
+    mut resp: serde_json::Value,
+    time: DateTime<Utc>,
+) -> Result<Vec<(FrostObs, FrostLatLonElev)>, Error> {
+    let ts_portion: &mut Vec<serde_json::Value> = resp
+        .get_mut("data")
+        .ok_or(Error::FindObs(
+            "couldn't find data field on root".to_string(),
+        ))?
+        .get_mut("tseries")
+        .ok_or(Error::FindObs(
+            "couldn't find tseries field on data".to_string(),
+        ))?
+        .as_array_mut()
+        .take()
+        .ok_or(Error::FindObs("couldn't get array of tseries".to_string()))?;
+
+    let observations = ts_portion
```

Comment:
Maybe this could be better named, since it contains both obs, and metadata?

Reply:
Ok if its called data?

Reply:
Yes, especially since that's what we put on the outside